### PR TITLE
Gesture controls

### PR DESCRIPTION
**Pull request by TheMatten**
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.

Current gesture controls work but look like an afterthought. This PR introduces circular design with progress bar and material icons that changes depending on current value.

Demo:
screenrecord-2018-08-15-12-00-51

Changed version seems to work without any problems.

### CODE REVIEW

**Comment by mauriciocolli** [comment]
Thanks for the pull request, looks good!

> Current gesture controls work but look like an afterthought

Yep, that's right, there's even a `TODO` there. I think you're talking about just the UI, but the code that handle gestures isn't really the best way to do it and it was meant as just a working solution, but yeah... (#503)

File: app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java [hunk 1]
```diff
@@ -992,12 +1042,31 @@ public boolean onScroll(MotionEvent e1, MotionEvent e2, float distanceX, float d
                 getWindow().setAttributes(lp);
                 if (DEBUG) Log.d(TAG, "onScroll().brightnessControl, currentBrightness = " + currentBrightness);
                 int brightnessNormalized = Math.round(currentBrightness * 100);
+                animateProgressBar(
+                        playerImpl.getBrightnessProgressBar(),
+                        (brightnessNormalized == 1 ? 0 : brightnessNormalized),
+                        100
+                );
+                if (brightnessNormalized < 25) {
+                    playerImpl.getBrightnessImageView().setImageDrawable(
```
Comment (mauriciocolli) [inline app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java]: Crash in old versions of Android.
A very simple solution, thankfully: use `AppCompatResources.getDrawable` instead.

I was also worried about performance, but [it seems to be cached](https://chromium.googlesource.com/android_tools/+/refs/heads/master/sdk/sources/android-25/android/support/v7/widget/AppCompatDrawableManager.java#180), both the Android and compat library versions, so it's all good.

File: app/src/main/java/org/schabi/newpipe/util/AnimationUtils.java [hunk 2]
```diff
@@ -43,6 +46,26 @@
...
     }
 
+    /**
+     * Animate the progress bar's progress
+     *
+     * @param progressBar progress bar that will be animated
+     * @param to          final progress value after animation
+     * @param duration    how long the animation will take, in milliseconds
+     */
+    public static void animateProgressBar(ProgressBar progressBar, int to, long duration) {
```
Comment (mauriciocolli) [inline app/src/main/java/org/schabi/newpipe/util/AnimationUtils.java]: Put this function below the others, but on top of the `Internals` block.

File: app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java (hunk 1 above)
Comment (TheMatten) [reply app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java]: Thanks for that - AS stopped complaining when I used it through getResources(), but yeah, I didn't check it manually...

File: app/src/main/java/org/schabi/newpipe/util/AnimationUtils.java (hunk 2 above)
Comment (TheMatten) [reply app/src/main/java/org/schabi/newpipe/util/AnimationUtils.java]: OK, I'll do

**Comment by TheMatten** [comment]
@mauriciocolli Design seemed like a biggest problem, but while tinkering with it, I thought about implementing some sort of speed-based behavior, so that user can change values more precisely by moving finger slowly. May try it, as well as maybe introduce some "fake" state of progress bar that rounds to the closest value after gesture, instead of that choppy movement.

**Comment by TheMatten** [comment]
@mauriciocolli While talking about gesture controls, there's one more thing that doesn't feel right - when using fast rewind controls, bright background may make icons barely visible:

np-rewind-nobg

E.g. YT web player solves this problem with dark circle as a background:
yt_circle_back_opt

The other thing we could do is temporarily dim the background (as with play/pause), but then rewind would still be in conflict with play button.

**Comment by TheMatten** [comment]
Wow - it seems like GestureDetector's and GestureDetectorCompat's onScroll gets called differently. Luckily, Compat version seems to be more sensitive (it works with slow movement).

After a lot of testing, to me the best solution seems to use event's distance directly as a value for progress incrementation and set max progress value based on screen size. Speed seems to be inconsistent and attempts to "normalize" value before use by e.g division or log result in number rounded down to zero when doing small movements.

I could maybe solve this by some overcomplicated solution I didn't think of or by skipping every Nth event and using progress animation again, but I don't think it would give us any advantage. Maybe if drawing on screen was too expensive, but actually, number of total motion events (including historical) should be <= touchscreen scan rate, which will probably match framerate (I've counted <= 60 per second on my device with `input swipe` command). And in reality, profiler barely shows any difference in load while using new implementation.

At the end, we can think of space used for gesture as an invisible seekbar, which is predictable, precise and smooth solution at the same time (and allows us to get from 0% to 100% at one long swipe). Value from it gets converted to nearest volume/brightness level and applied.

Demo:
screenrecord-2018-08-20-09-41-22

**Comment by TheMatten** [comment]
(Sorry for force push - I reworded one commit message)

**Comment by mauriciocolli** [comment]
> when using fast rewind controls, bright background may make icons barely visible

IMO it doesn't matter much, the user is focused on the video and will certainly notice the rewind/fast forward. Though I don't mind if you want to make more visible, feel free to do it (of course, in a way that doesn't draw too much attention).

And the fact that the icons stay very shortly in screen (video is already loaded) makes the player seems snappy, though the "artifacts" (progress bar visible in a split second, for example) can be seen as a buggy player UI to the user. The optimal solution would be to keep the short animation and fix them.

> we can think of space used for gesture as an invisible seekbar

Exactly, I like this implementation, works and feels much better.

---

Oh, and if you don't mind, please, when this PR is ready remove the commits that cancel each other.

File: app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java [hunk 3]
```diff
@@ -392,6 +402,18 @@ public void onPlaybackParameterChanged(float playbackTempo, float playbackPitch,
         private RelativeLayout windowRootLayout;
         private View secondaryControls;
 
+        private int maxGestureLength;
+
+        {
+            DisplayMetrics m = getResources().getDisplayMetrics();
```
Comment (mauriciocolli) [inline app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java]: What do you think of using the `rootView`'s width and height?

The code would be simpler, and using this approach, we could already support the multi window mode just by adding a listener to the `rootView` layout changes, that when we detect one, we could simply update the variable and the gesture would work correctly in the new size.

File: app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java [hunk 4]
```diff
@@ -930,21 +979,9 @@ public boolean onDown(MotionEvent e) {
...
-        private final String brightnessUnicode = new String(Character.toChars(0x2600));
-        private final String volumeUnicode = new String(Character.toChars(0x1F508));
+        private int maxVolume = playerImpl.getAudioReactor().getMaxVolume();
 
         private final int MOVEMENT_THRESHOLD = 40;
-        private final int eventsThreshold = 8;
-        private boolean triggered = false;
-        private int eventsNum;
 
         // TODO: Improve video gesture controls
```
Comment (mauriciocolli) [inline app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java]: I think we can remove this `TODO`.

File: app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java (hunk 3 above)
Comment (TheMatten) [reply app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java]: Okay, will do.

File: app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java (hunk 4 above)
Comment (TheMatten) [reply app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java]: Great :D

**Comment by TheMatten** [comment]
@mauriciocolli 

> Though I don't mind if you want to make more visible, feel free to do it (of course, in a way that doesn't draw too much attention).

Just a simple background similar to YT or those new gesture controls to make it consistent.

> And the fact that the icons stay very shortly in screen (video is already loaded) makes the player seems snappy, though the "artifacts" (progress bar visible in a split second, for example) can be seen as a buggy player UI to the user.

Otherwise it's fine, I was just interested in "visibility". Frankly, any sort of long animation or effects would just get in the way of users while watching.

Oh, one more thing - #1603 (3. point) and #1199 - those raster images look terrible when scaling, but that's thing for separate PR.

**Comment by TheMatten** [comment]
Okay, cleaned up history - commits changing gesture control are squashed into one big.

File: app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java [hunk 5]
```diff
@@ -424,6 +440,20 @@ public void initViews(View rootView) {
             titleTextView.setSelected(true);
             channelTextView.setSelected(true);
 
+            getRootView().addOnLayoutChangeListener((view, l, t, r, b, ol, ot, or, ob) -> {
+                // Use smaller value to be consistent between screen orientations
```
Comment (mauriciocolli) [inline app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java]: Little thing that's bothering the micro-optimization part of me, we could return if the size itself didn't changed:
```java
if (left == oldLeft && right == oldRight) return;
```

Comment (TheMatten) [reply app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java]: What about `if (l != ol || t != ot || r != or || b != ob) { // do things... }` ?

**Comment by TheMatten** [comment]
Added check to max length updates - I should be ready for merging.

File: app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java (hunk 5 above)
Comment (mauriciocolli) [reply app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java]: It's fine, this piece of code is short enough to be a good a choice.

File: app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java [hunk 6]
```diff
@@ -956,63 +993,90 @@ public boolean onScroll(MotionEvent e1, MotionEvent e2, float distanceX, float d
...
+                float currentProgressPercent =
+                        (float) playerImpl.getBrightnessProgressBar().getProgress() / playerImpl.getMaxGestureLength();
+                WindowManager.LayoutParams layoutParams = getWindow().getAttributes();
+                layoutParams.screenBrightness = currentProgressPercent;
+                getWindow().setAttributes(layoutParams);
+
+                if (DEBUG) Log.d(TAG, "onScroll().brightnessControl, currentBrightness = " + currentProgressPercent);
+
+                if (currentProgressPercent < 0.25) {
+                    playerImpl.getBrightnessImageView().setImageDrawable(
```
Comment (mauriciocolli) [inline app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java]: I think we could improve readibility and be less redudant here (and above) introducing a variable:
```java
// We could initialize at the "else" value, but I think
// this way makes it more readable.
final int iconRes; 
if (currentProgressPercent < 0.25) {
    iconRes = R.drawable.ic_brightness_low_white_72dp;
} else if (currentProgressPercent < 0.75) {
    iconRes = R.drawable.ic_brightness_medium_white_72dp;
} else {
    iconRes = R.drawable.ic_brightness_high_white_72dp;
} 
```

What do you think?

File: app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java [hunk 7]
```diff
@@ -424,6 +440,20 @@ public void initViews(View rootView) {
             titleTextView.setSelected(true);
             channelTextView.setSelected(true);
 
+            getRootView().addOnLayoutChangeListener((view, l, t, r, b, ol, ot, or, ob) -> {
+                if (l != ol || t != ot || r != or || b != ob) {
+                    // Use smaller value to be consistent between screen orientations
+                    // (and to make usage easier)
+                    int width = r - l, height = b - t;
+                    maxGestureLength = (int) ((width > height ? width : height) * MAX_GESTURE_LENGTH);
```
[the remark this answers is not in the document]
Comment (TheMatten) [reply app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java]: Doesn't it make more sense like this? We're doing update when something changes, not returning when values stayed the same...

File: app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java (hunk 6 above)
Comment (TheMatten) [reply app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java]: Not a bad idea, will do.

Comment (TheMatten) [reply app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java]: Hmm, ternary operator is right-associative... :
```
playerImpl.getVolumeImageView().setImageDrawable(
        AppCompatResources.getDrawable(getApplicationContext(),
                currentProgressPercent <= 0
                        ? R.drawable.ic_volume_off_white_72dp
                        : currentProgressPercent < 0.25
                        ? R.drawable.ic_volume_mute_white_72dp
                        : currentProgressPercent < 0.75
                        ? R.drawable.ic_volume_down_white_72dp
                        : R.drawable.ic_volume_up_white_72dp
        )
);
```
Can I use this? :D

Comment (mauriciocolli) [reply app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java]: Definitely hard to read, but if you keep the condition result in the same line, yeah. A variable here can help it again to keep the line shorter and the style closer to what is used currently:

```java
final int resId = currentProgressPercent <= 0 ? R.drawable.ic_volume_off_white_72dp
        : currentProgressPercent < 0.25 ? R.drawable.ic_volume_mute_white_72dp
        : currentProgressPercent < 0.75 ? R.drawable.ic_volume_down_white_72dp
        : R.drawable.ic_volume_up_white_72dp;
```

File: app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (mauriciocolli) [reply app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java]: Just throw a `Math.min(width, height)` in there, even easier to read.

[the remark this answers is not in the document]
Comment (TheMatten) [reply app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java]: Much better idea, thanks.

File: app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java (hunk 6 above)
Comment (TheMatten) [reply app/src/main/java/org/schabi/newpipe/player/MainVideoPlayer.java]: I just wanted to avoid additional variable... but okay, will do.

**Comment by TheMatten** [comment]
Fixed listener and res selection.

**Comment by TheMatten** [comment]
Great! Thanks for your help (and patience) :)

**Pull request merged**